### PR TITLE
Implements Easy To Borrow Reality Modeling (WIP)

### PR DESCRIPTION
**Pull request by gsalaz98**
### This commit is a work in progress, pushed for review

Implements easy to borrow/short restrictions for algorithms.
#### Related Issue

#5027

#### Motivation and Context

Reality modeling for improved backtesting and live trading

#### Requires Documentation Change

Yes, should be documented somewhere

#### How Has This Been Tested?

New tests (pending)

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.CSharp/ShortableProviderOrdersRejectedRegressionAlgorithm.cs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                    throw new Exception($"Expected market order of -1001 BAC to not fail");
+                }
+
+                _invalidatedNewOrderWithPortfolioHoldings = true;
+            }
+        }
+
+        private void HandleOrder(OrderTicket orderTicket)
+        {
+            if (orderTicket.SubmitRequest.Status == OrderRequestStatus.Error)
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/ShortableProviderOrdersRejectedRegressionAlgorithm.cs:129]: There should be an error message of why it errored, let's assert on it too 👍

File: Algorithm.CSharp/ShortableProviderOrdersRejectedRegressionAlgorithm.cs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            {
+                _ordersDenied.Add(orderTicket);
+                return;
+            }
+
+            _ordersAllowed.Add(orderTicket);
+        }
+
+        public class RegressionTestShortableProvider : IShortableProvider
```
Comment (Martin-Molinero) [inline Algorithm.CSharp/ShortableProviderOrdersRejectedRegressionAlgorithm.cs:138]: Some for this Shortable provider, let's use the real file-based mockup data at `shortable/testbrokerage`

File: Common/Securities/SecurityTransactionManager.cs [hunk 3]
```diff
@@ -260,6 +260,27 @@ public IEnumerable<OrderTicket> GetOpenOrderTickets(Func<OrderTicket, bool> filt
...
 
+        /// <summary>
+        /// Gets the remaining quantity to be filled from open orders, i.e. order size minus quantity filled
+        /// </summary>
+        /// <param name="filter">Filters the order tickets to be included in the aggregate quantity remaining to be filled</param>
+        /// <returns>Total quantity that hasn't been filled yet for all orders that were not filtered</returns>
+        public decimal GetOpenOrdersRemainingQuantity(Func<OrderTicket, bool> filter = null)
+        {
+            return GetOpenOrderTickets(filter)
+                .Aggregate(0m, (d, t) => d + t.Quantity - t.QuantityFilled);
```
Comment (Martin-Molinero) [inline Common/Securities/SecurityTransactionManager.cs:271]: Believe `orderSizing` should use this new method?

**Review by Martin-Molinero — COMMENTED** [review]
Nice progress! Leaving a few suggestions

**Review by jaredbroad — APPROVED** [review]

**Pull request merged**